File: http/operation/workflow/put.js
/**
 * Creation Date: 22/05/2019
 * Summary: PUT method of HTTP
 * Description: To PUT data from a URL using HTTP Connector
 */

//Starting execution of flintbit
log.trace("Started executing 'flint-util:http:operation:workflow:put.js'");

//method
method = "put";

input_clone = JSON.parse(input); //For checking purposes

//Connector name - mandatory
connector_name = "http";
connector_call = call.connector(connector_name);
log.info("Connector Name: "+connector_name);
/*
if(input_clone.hasOwnProperty("connector_name")){
    connector_name = input.get("connector_name");
    if(connector_name!=null || connector_name!=""){
        connector_call = call.connector(connector_name);
        log.info("Connector Name: "+connector_name);
    }
    else{
        log.error("Connector Name is null or empty string.");
    }
}
else{
    log.error("'connector_name' key not given in input JSON");
}
*/
//URL - mandatory
if(input_clone.hasOwnProperty("url")){
    url = input.get("url");
    if(url!=null || url!=""){
        connector_call.set("url",url);
        log.info("URL: "+url);
    }
    else{
        log.error("URL is null or empty string.");
    }
}
else{
    log.error("'url' key not given in input JSON");
}

//Body - mandatory
if(input_clone.hasOwnProperty("body")){
    body = input.get("body");
    if(body!=null || body!=""){
        connector_call.set("body",body);
        log.info("Body: "+body);
    }
    else{
        log.error("Body is null or empty string.");
    }
}
else{
    log.error("'body' key not given in input JSON");
}

//Headers - not mandatory
if(input_clone.hasOwnProperty("headers")){
    headers = input.get("headers");
    if(headers!=null || headers!=""){
        connector_call.set("headers",headers);
        log.info("Headers: "+headers);
    }
}

//Is Proxy - not mandatory
if(input_clone.hasOwnProperty("is_proxy")){
    is_proxy = input.get("is_proxy");
    if(is_proxy!=null || is_proxy!=""){
        connector_call.set("is-proxy",is_proxy);
        log.info("Is-Proxy: "+is_proxy);
    }
}

//Proxy - not mandatory
if(input_clone.hasOwnProperty("proxy")){
    proxy = input.get("proxy");
    if(proxy!=null || proxy!=""){
        connector_call.set("proxy",proxy);
        log.info("Proxy: "+proxy);
    }
}

//Timeout - not mandatory
if(input_clone.hasOwnProperty("timeout")){
    timeout = input.get("timeout");
    if(timeout!=null || timeout!=""){
        timeout = parseInt(timeout);
    }
    else{
        timeout = 60000;
    }
}
else{
    timeout = 60000;
}
connector_call.set("timeout",timeout);
log.info("Timeout: "+timeout);

//Calling the HTTP connector
response = connector_call.set("method",method).sync();

//Response's meta parameters
response_exitcode = response.exitcode();
response_message = response.message();

//Response's Body
response_body = response.get("body");

//Response's Headers
response_headers = response.get("headers");

if(response_exitcode==0){
    log.info("Successfull execution of method:"+method);
    log.info("Method result:"+response_body);
    //User Message
    user_message = "Response-Body: "+response_body+"  |  Response-Headers: "+response_headers;
    output.set("exit-code",0).set("result",response_body).set("user_message",user_message);
    log.trace("Finished executing 'flint-util:http:operation:workflow:put.js' successfully")
}

else{
    log.error("Failure in execution of method:"+method);
    output.set("error",response_message).set("exit-code",response_exitcode);
    log.trace("Finished executing 'flint-util:http:operation:workflow:put.js' with errors");
}

File: http/operation/put.rb
=begin
#
#  INFIVERVE TECHNOLOGIES PTE LIMITED CONFIDENTIAL
#  _______________________________________________
# 
#  (C) INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE
#  All Rights Reserved.
#  Product / Project: Flint IT Automation Platform
#  NOTICE:  All information contained herein is, and remains
#  the property of INFIVERVE TECHNOLOGIES PTE LIMITED.
#  The intellectual and technical concepts contained
#  herein are proprietary to INFIVERVE TECHNOLOGIES PTE LIMITED.
#  Dissemination of this information or any form of reproduction of this material
#  is strictly forbidden unless prior written permission is obtained
#  from INFIVERVE TECHNOLOGIES PTE LIMITED, SINGAPORE.
=end

# begin
@log.trace("Started executing 'flint-util:http:operation:put.rb' flintbit...")
begin
    # Flintbit Input Parameters
    # mandatory
    @connector_name = @input.get('connector_name')   # Name of the HTTP Connector
    @request_method = 'put'           # HTTP Request Method
    @request_url = @input.get('url')                 # HTTP Request URL
    @request_body = @input.get('body')               # HTTP Request Body
    @request_headers = @input.get('headers')         # HTTP Request Headers
    @is_proxy=@input.get('is-proxy')		     #proxy (true or false)
    @proxy_details=@input.get('proxy')		     #if proxy is true then provide details of the proxy i. hostname,port,protocol
    # optional
    @request_timeout = @input.get('timeout') # Timeout in milliseconds, taken by the Connetor to serve the request

    @log.info("Flintbit input parameters are, connector name :: #{@connector_name} | url :: #{@request_url} | method :: #{@request_method} |
    headers :: #{@request_headers} | body :: #{@request_body} | timeout :: #{@request_timeout}")

    if @is_proxy.nil?		
           @is_proxy = false
    end

    @log.trace('Calling HTTP Connector...')
    if @request_timeout.to_s.empty?
        response = @call.connector(@connector_name)
                        .set('method', @request_method)
                        .set('url', @request_url)
                        .set('body', @request_body)
			.set('is-proxy',@is_proxy)
			.set('proxy',@proxy_details)
                        .set('headers', @request_headers)
                        .sync
    else
        response = @call.connector(@connector_name)
                        .set('method', @request_method)
                        .set('url', @request_url)
                        .set('body', @request_body)
                        .set('headers', @request_headers)
			.set('is-proxy',@is_proxy)
			.set('proxy',@proxy_details)
                        .set('timeout', @request_timeout)
                        .sync
    end

    # HTTP Connector Response Meta Parameters
    response_exitcode = response.exitcode  # Exit status code
    response_message = response.message    # Execution status messages

    # HTTP Connector Response Parameters
    @response_body = response.get('body')       # Response Body
    @response_headers = response.get('headers') # Response Headers

    if response.exitcode == 0
        @log.info("Success in executing HTTP Connector where, exitcode :: #{response_exitcode} | message :: #{response_message}")
        @log.info("HTTP Response Headers :: #{@response_headers} | HTTP Response Body :: #{@response_body}")
        @output.set('result', @response_body)
        @log.trace("Finished executing 'put' flintbit with success...")
    else
        @log.error("Failure executing HTTP Connector where, exitcode :: #{response_exitcode} | message :: #{response_message}")
        @output.set('error', response_message)
        @log.trace("Finished executing 'put' flintbit with error...")
    end
rescue Exception => e
    @log.error(e.message)
    @output.set('exit-code', 1).set('message', e.message)
end
# end
